python_modules/libraries/dagster-aws/dagster_aws/athena/resources.py

```diff
@@ -0,0 +1,262 @@
+import csv
+import io
+import os
+import uuid
+
+try:
+    from urlparse import urlparse
+except ImportError:
+    from urllib.parse import urlparse
+
+import boto3
+from botocore.stub import Stubber
+from dagster import Field, StringSource, check, resource
+
+
+class AthenaError(Exception):
+    pass
+
+
+class AthenaTimeout(AthenaError):
+    pass
+
+
+class AthenaResource:
+    def __init__(self, client, workgroup="primary", polling_interval=5, max_polls=120):
+        self.client = client
+        self.workgroup = workgroup
+        self.max_polls = max_polls
+        self.polling_interval = polling_interval
+
+    def execute_query(self, query, fetch_results=False):
+        """Synchronously execute a single query against Athena. Will return a list of rows, where
+        each row is a tuple of stringified values, e.g. SELECT 1 will return [("1",)].
+
+        Args:
+            query (str): The query to execute.
+            fetch_results (Optional[bool]): Whether to return the results of executing the query.
+                Defaults to False, in which case the query will be executed without retrieving the
+                results.
+
+        Returns:
+            Optional[List[Tuple[Optional[str], ...]]]: Results of the query, as a list of tuples,
+                when fetch_results is set. Otherwise, return None. All items in the tuple are
+                represented as strings except for empty columns which are represented as None.
+        """
+        check.str_param(query, "query")
+        check.bool_param(fetch_results, "fetch_results")
+        execution_id = self.client.start_query_execution(
+            QueryString=query, WorkGroup=self.workgroup
+        )["QueryExecutionId"]
+        self._poll(execution_id)
+        if fetch_results:
+            return self._results(execution_id)
+
+    def _poll(self, execution_id):
+        retries = self.max_polls
+        state = "QUEUED"
+
+        while retries >= 0 and state in ["QUEUED", "RUNNING"]:
+            execution = self.client.get_query_execution(QueryExecutionId=execution_id)[
+                "QueryExecution"
+            ]
+            state = execution["Status"]["State"]
+            if self.max_polls > 0:
+                retries -= 1
+
+        if retries < 0:
+            raise AthenaTimeout()
+
+        if state != "SUCCEEDED":
+            raise AthenaError(execution["Status"]["StateChangeReason"])
+
+    def _results(self, execution_id):
```

Comment:
This is a potentially controversial decision: I chose to return results from Athena's [output CSV](https://docs.aws.amazon.com/athena/latest/ug/querying.html#querying-identifying-output-files) instead of Athena's [GetQueryResults endpoint](https://docs.aws.amazon.com/athena/latest/APIReference/API_GetQueryResults.html).

I believe this makes for a simpler implementation and easier testing, but it also means that all non-null columns are cast to strings 😬.

If we think it's critical that we preserve types, just say the word and I can make that change. It does mean the PR will get much larger though to account for [the various complex types](https://docs.aws.amazon.com/athena/latest/ug/data-types.html) that Athena queries support.